File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WSBodyWritables.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import akka.stream.scaladsl.Source
import akka.util.ByteString
import play.api.mvc.MultipartFormData
import play.core.formatters.Multipart

/**
 * JSON, XML and Multipart Form Data Writables used for Play-WS bodies.
 */
trait WSBodyWritables extends DefaultBodyWritables with JsonBodyWritables with XMLBodyWritables {

  implicit val bodyWritableOf_Multipart: BodyWritable[Source[MultipartFormData.Part[Source[ByteString, _]], _]] = {
    val boundary = Multipart.randomBoundary()
    val contentType = s"multipart/form-data; boundary=$boundary"
    BodyWritable(b => SourceBody(Multipart.transform(b, boundary)), contentType)
  }

}

object WSBodyWritables extends WSBodyWritables

File: framework/src/play-ws/src/main/java/play/libs/ws/WSBodyWritables.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import akka.stream.javadsl.Source;
import akka.util.ByteString;
import play.mvc.Http;
import play.mvc.MultipartFormatter;

/**
 * JSON, XML and Multipart Form Data Writables used for Play-WS bodies.
 */
public interface WSBodyWritables extends DefaultBodyWritables, XMLBodyWritables, JsonBodyWritables {

    default SourceBodyWritable multipartBody(Source<? super Http.MultipartFormData.Part<Source<ByteString, ?>>, ?> body) {
        String boundary = MultipartFormatter.randomBoundary();
        Source<ByteString, ?> source = MultipartFormatter.transform(body, boundary);
        String contentType = "multipart/form-data; boundary=" + boundary;
        return new SourceBodyWritable(source, contentType);
    }

}
